@uppy/companion: add parent folder path in onedrive provider

JammingBen opened the pull request:
Adds the parent folder path to the OneDrive provider. This can be used to determine and distinguish different source locations of files.

Relates to https://github.com/transloadit/uppy/issues/4034 (although it doesn't fix it completely because other providers are still missing).

arturi commented:
Thank you! Note that we have `relativePath` meta data key in Uppy, I wonder if it should be used here instead of `parentPath`.

https://github.com/transloadit/uppy/blob/72f579e21690b7d7180cc9e424ab497c35ef3bdf/packages/%40uppy/dashboard/src/Dashboard.jsx#L386-L390

mifi commented on packages/@uppy/companion/src/server/provider/onedrive/adapter.js:
```diff
@@ -39,6 +39,10 @@ const getItemId = (item) => {
 }
 
+const getParentPath = (item) => {
+  return item.parentReference ? item.parentReference.path : null
```
```suggestion
  return item.parentReference?.path
```

mifi reviewed (commented):
what's a 'flat' file? and what does relativePath mean in the context of onedrive? I believe relativePath is only set for local files in Uppy? (not remote/provider files).

Also I'm not sure what's the difference between a file's relativePath and a file's path - what is a path relative to? should it instead be called path?

The way I understand it, parentPath in OneDrive is the whole absolute path to a file excluding the file name, kind of like [dirname()](https://nodejs.org/api/path.html#pathdirnamepath) in node.js, am i right?

JammingBen commented:
> what's a 'flat' file?

By that I meant files that are not located inside another directory but in the root directory.

> The way I understand it, parentPath in OneDrive is the whole absolute path to a file excluding the file name, kind of like [dirname()](https://nodejs.org/api/path.html#pathdirnamepath) in node.js, am i right?

Yes, that's my understanding as well. Hence I figured `parentPath` would be a fitting name. Something like `directoryPath` would work as well I guess.

> Also I'm not sure what's the difference between a file's relativePath and a file's path - what is a path relative to? should it instead be called path?

In that specific case there is no difference I believe. `relativePath` makes sense for local files as it's the path relative to the folder the file is located in on the OS. However, here, all paths are always relative to the OneDrive root = the file path.

mifi commented:
Ok, I believe relativePath is something completely different than this then (it also includes the file's name), so we shouldn't mix the two together.

IMO something like `directoryPath` is a better name than `parentPath`, but we can let others weigh in too. @arturi ?

I find the root path `/drive/root:` a bit odd with the trailing colon. Am I right that if someone selects a file that has the (unix equivalent) path `/mydir/myfile.pdf`, then `item.parentReference?.path` would become `/drive/root:/mydir` ?

mifi commented:
related? #4423

arturi commented:
Also related: https://github.com/transloadit/uppy/issues/4034?

arturi commented:
Also thinking that `/root:` is a bit weird, especially if only GoogleDrive-specific, so maybe the path should be `/googledrive/myfolder/file.txt` or `/googledrive/file.txt`.

Murderlon commented:
My two cents:

- Name it `directoryPath`
- Make the difference clear between `requestPath` and `directoryPath`. Comment could be enough.
- Remove `/root:` from the path, if possible.

arturi commented:
Note: we add `relativePath` to `file.id`, so that two identical files from different folders are allowed in Uppy. Maybe we should account for that too with the new property.

JammingBen commented:
Okay so I renamed the property to `directoryPath`, added comments and strip `/drive/root:` (which sounds reasonable to me).

It relates to https://github.com/transloadit/uppy/issues/4423 and https://github.com/transloadit/uppy/issues/4034 although it does not fix them because this is for OneDrive only.

mifi commented on packages/@uppy/companion/src/server/provider/onedrive/adapter.js:
```diff
@@ -40,13 +40,19 @@ const getItemId = (item) => {
 
+const getDirectoryPath = (item) => {
+  // absolute path to the directory the item is located in
+  return item.parentReference?.path?.replace('/drive/root:', '')
```
```suggestion
  const strippedPath = item.parentReference?.path?.replace(/^\/drive\/root:/, '')
  return strippedPath === '' ? '/' : strippedPath
```

mifi commented on packages/@uppy/companion/test/fixtures/onedrive.js:
```diff
@@ -2,4 +2,5 @@ const defaults = require('./constants')
 
 module.exports.expects = {
   itemRequestPath: `${defaults.ITEM_ID}?driveId=DUMMY-DRIVE-ID`,
+  directoryPath: '',
```
```suggestion
  directoryPath: '/',
```

mifi reviewed (changes requested):
I think the root path should be `/`. also there's a bug where it doesn't match from the start of the string

dschmidt commented:
I've been playing around further with this with @JammingBen and we realized that this is not all we need.

Basically we want to achieve that recursive cloud uploads work - relative to the selected folder, just like local folders 
The second change we (probably) would like to make is to add a `relativePath` property to the file via the `ProviderView` here: https://github.com/transloadit/uppy/blob/2b9958c4f8faee85b3c922555efe2c0109d7a2fb/packages/%40uppy/provider-views/src/ProviderView/ProviderView.jsx#L277-L292

as the server has no knowledge about the selection. We would be doing something along these lines:
```javascript              
             if(!newFile.relativePath) {
                const relativePath = (newFile.directoryPath.replace(file.directoryPath, '') || '/') + newFile.name
                newFile.relativePath = relativePath
              }
``` 

The `relativePath` for local upload files contains the path of filename relative to the selected folder, so it makes full sense to reuse that property name here if we make it contain the path including the filename (imho)

If we do that, we might as well just add the filename in the server and name the property `absolutePath` instead of `directoryPath`.
To me that seems a tad more consistent and feels a little less awkward than `directoryPath` - what do you think?

Would you like a separate PR for the `ProviderView` change or would you prefer us to add it to this one?

dschmidt commented:
When we agree on something, we should add it to the documentation here: https://github.com/transloadit/uppy.io/blob/4bc81c342bdf48fd9083c6b8975f1f3b078d4580/docs/companion.md?plain=1#L818

mifi commented:
I've been doing some more thinking around this, and I'm not sure if this is the way to do it, because not all providers support returning the path to Companion in their APIs, so it would't be consistent between providers. We probably want to implement a solution that work across all (most) providers.

Will continue the discussion here for now: #4034 (because it is not onedrive specific)

dschmidt commented:
@JammingBen I guess you can close this one, as a different solution is being worked on and we use the same approach in our fork already

Murderlon commented:
Superseded by #4537
